src/main/kotlin/org/rust/lang/core/psi/RsPsiImplUtil.kt

```diff
@@ -28,5 +30,86 @@ object RsPsiImplUtil {
+
+        return null
+    }
+
+    /**
+     * @return null if default scope should be used
+     */
+    fun getDeclarationUseScope(element: RsVisible): SearchScope? =
+        getDeclarationUseScope(element, RsVisibility.Public)
+
+    private fun getDeclarationUseScope(
+        element: RsVisible,
+        restrictedVis: RsVisibility
+    ): SearchScope? {
+        when (element) {
+            is RsEnumVariant -> return getDeclarationUseScope(element.parentEnum, restrictedVis)
+            is RsFieldDecl -> return getDeclarationUseScope(
+                element.contextStrict<RsFieldsOwner>() as RsVisible,
+                element.visibility
+            )
+        }
+
+        val owner = PsiTreeUtil.getContextOfType(
+            element,
+            true,
+            RsItemElement::class.java,
+            RsMod::class.java
+        ) as? RsElement ?: return null
+
+        return when (owner) {
+            // Trait members inherit visibility from the trait
+            is RsTraitItem -> getDeclarationUseScope(owner)
+
+            is RsImplItem -> {
+                // Members of `impl Trait for ...` inherit visibility from the implemented trait
+                val traitRef = owner.traitRef
+                if (traitRef != null) return getDeclarationUseScope(traitRef.resolveToTrait ?: return null)
+
+                // Inherent impl members
+                getTopLevelDeclarationUseScope(element, owner.containingMod, restrictedVis)
+            }
+
+            is RsMod -> getTopLevelDeclarationUseScope(element, owner, restrictedVis)
+            is RsForeignModItem -> getTopLevelDeclarationUseScope(element, owner.containingMod, restrictedVis)
+
+            // In this case `owner` is function or code block, i.e. it's local scope
+            else -> LocalSearchScope(owner)
+        }
+    }
+
+    private fun getTopLevelDeclarationUseScope(
+        element: RsVisible,
+        containingMod: RsMod,
+        restrictedVis: RsVisibility
+    ): SearchScope? {
+        val restrictedMod = when (val visibility = restrictedVis.intersect(element.visibility)) {
+            RsVisibility.Public -> return null
+            RsVisibility.Private -> containingMod
+            is RsVisibility.Restricted -> visibility.inMod
+        }
+
+        if (!restrictedMod.hasChildModules()) return LocalSearchScope(containingMod)
+
+        val virtualFile = restrictedMod.contextualFile.virtualFile ?: return null
+
+        // Use parent directory (with subdirectories) as a scope. We assume that all child modules are located in
+        // the same directory. The assumption is generally wrong: a file from any directory can be included to
+        // the module structure via `#[path]` attribute or `include!()` macro. But it works in the most cases.
+        // Maybe we should calculate more precise scope by traversing module tree.
+        return GlobalSearchScopesCore.DirectoryScope(restrictedMod.project, virtualFile.parent, true)
```

Comment:
We don't support `include` macro at this moment so it's ok
But `#[path]` attribute is not so rare so we can break some code that works now. How much does it affect `find usages` performance if we use default search scope when `restrictedMod` has child modules?

Looks like there are three main cases:
1. item is public and we should provide default scope (it's already supported)
2. item is private and containing mod doesn't have child modules
3. item is crate private

`1` will not be changed
`2` will not be changed
`3` not sure how much it affects performance

Reply:
There is also 4th item: item is private and containing mod have child modules. This is enough usual case, I guess, and we optimize it significantly.

About 3th item, it it reasonable when searching usages of some item inside a dependency crate, or in project with mane crates, like `rustc`.

> But #[path] attribute is not so rare so we can break some code that works now.

`#[path]` attribute - yes, but `#[path]` attribute that refer to a file from parent directory - I'm not sure.

Reply:
Oh, I found a lot of usages of `path` attribute like `#[path = ".../something"]` in stdlib! Ok, looks like it's not a rare case. For now I've replaced it with `return null` (use default scope). I future I'll try to traverse module structure to calculate more precise scope